test: add test for error codes doc and impl

ChALkeR opened the pull request:
This adds several sanity checks for error codes, ref: #21440.
An early version of this test was used to find #21440 (and is something I came up after seeing #21421).

It scans:
1. all natives (js sources),
2. `doc/api/*.md` documentation
3. `src/node_errors.h` (errors definition from the C++ side).

There is also a whitelist of manually created errors from JS side, currently consisting of `ERR_HTTP2_ERROR` and `ERR_UNKNOWN_BUILTIN_MODULE`. That could be improved, I guess.

The performed checks:
1. All errors used from JS should be defined in `internal/errors` and present in its `.codes` object.
    Whitelist with two exceptions of manually created errors (listed above) applies.
2. All errors instantiated from JS without arguments should support 0-arguments version.
    As in `new ERR_SOMETHING()`, ref: #21421.
3. All errors mentioned in doc should defined either in JS (`internal/errors`), C++ (`src/node_errors.h`), or in the whitelist with two manually created errors.
4. All errors mentioned anywhere should be documented.
5. Documentation of error codes should be sorted, have no repeats, and include exactly one entry formatted as `/\n### (ERR_[A-Z0-9_]+)\n` for every error code mentioned in the documentation.
6. All doc entries for error codes should have appropriate anchors.

The checks are a bit hacky (and scan the source code as text), but they work and I would prefer keeping it simple as we have a common markup for all those.

There is also a `--report` flag, which prints all the current issues and exits without asserting, for manual inspection.

Current output on v10.5.0:
```console
$ node --expose-internals test/parallel/test-error-codes-impl-doc.js --report
Report mode
js: ERR_TLS_RENEGOTIATE - missing JS implementation (source)
js: ERR_MISSING_DYNAMIC_INTSTANTIATE_HOOK - missing JS implementation (source)
js: ERR_HTTP2_SETTINGS_CANCEL - missing JS implementation (source)
init: ERR_METHOD_NOT_IMPLEMENTED - failed 0 args init, but is called with "()"
impl: ERR_STREAM_READ_NOT_IMPLEMENTED - missing implementation, documented
impl: ERR_VALUE_OUT_OF_RANGE - missing implementation, documented
impl: ERR_HTTP2_STREAM_CLOSED - missing implementation, mentioned in doc/api/
doc: ERR_UNKNOWN_BUILTIN_MODULE - missing documentation
doc: ERR_TLS_RENEGOTIATE - missing documentation
doc: ERR_MISSING_DYNAMIC_INTSTANTIATE_HOOK - missing documentation
doc: ERR_HTTP2_SETTINGS_CANCEL - missing documentation
doc: ERR_HTTP2_ERROR - missing documentation
doc: ERR_TRANSFERRING_EXTERNALIZED_SHAREDARRAYBUFFER - missing documentation
doc: ERR_HTTP2_STREAM_CLOSED - missing documentation
doc: ERR_HTTP2_HEADERS_AFTER_RESPOND - is out of order
doc: ERR_STREAM_DESTROYED - is out of order
doc: ERR_TLS_RENEGOTIATION_DISABLED - is out of order
There were 17 problems found
```

I believe that all those are actual problems that need to be fixed.

This test currently fails and is blocked on #21440, i.e. fixing all those errors.

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

ChALkeR commented:
/cc @nodejs/tsc this is a bit hacky test, so I would be happy with some extra feedback on that.
I think it's valuable though (as it found some problems) and imo it should be relatively stable.

joyeecheung commented on test/parallel/test-error-codes-impl-doc.js:
```diff
@@ -0,0 +1,104 @@
+// File containing error definitions
+const jsdata = natives['internal/errors'];
+// File containing error documentation
+const docdata = fs.readFileSync('doc/api/errors.md', 'utf-8');
```
Can you make this an absolute path so that it can be run somewhere other than the project directory?

ChALkeR replied:
Fixed, thanks @joyeecheung!

joyeecheung commented:
I like the idea, a few notes:

> All errors mentioned in doc (doc/api/errors.md) should defined either in JS (internal/errors), C++ (src/node_errors.h), or in the whitelist with two manually created errors.

This does not apply to errors that are removed (those are the `missing implementation` ones). We did remove errors and merge them, but the documentation for those error codes need to stay forever.

jasnell commented:
Great idea :-)

ChALkeR commented:
@joyeecheung @jasnell But unused errors were removed several times from documentation.
E.g.:
 * 186857f15c — removes `ERR_INVALID_ARRAY_LENGTH`
 * 564048dc29 — removes `ERR_INVALID_DOMAIN_NAME` 
 * 362694401f — renames `ERR_STRING_TOO_LARGE`
 * 301f6cc553 — removes `ERR_FS_WATCHER_ALREADY_STARTED` and `ERR_FS_WATCHER_NOT_STARTED`
 * 5e3f51648e
 * 6e1c25c456
 * a82b1b7b30
 * 0babd181a0
 * e9358af581
 * 1cdb41f287
 * e6b69b9418
 * 1b54371c50
 * 8ca9338655

Do we want to restore all those?

ChALkeR commented:
I added a doc sorting/formatting check (see above, I updated the post), this is what it currently prints:
```console
doc: ERR_HTTP2_ALTSVC_INVALID_ORIGIN - is out of order
doc: ERR_HTTP2_HEADERS_AFTER_RESPOND - is out of order
doc: ERR_STREAM_DESTROYED - is out of order
doc: ERR_TLS_RENEGOTIATION_DISABLED - is out of order
```

ChALkeR commented:
One more check added: all doc entries should have appropriate anchors.
`ERR_SCRIPT_EXECUTION_TIMEOUT` doesn't:
```console
doc: ERR_SCRIPT_EXECUTION_TIMEOUT - missing anchor or not properly formatted
```

ChALkeR commented:
Updated with a scan of all `doc/api/*.md` files, found this:
```console
impl: ERR_HTTP2_STREAM_CLOSED - missing implementation, mentioned in doc/api/
```

ChALkeR commented:
I filed #21483, #21484, #21485, #21487 to fix some of those issues, but the remaining ones still need to be fixed.

vsemozhetbyt commented on test/parallel/test-error-codes-impl-doc.js:
```diff
@@ -0,0 +1,156 @@
+  doc: new Set(), // mentioned in doc/api/*.md
+  cpp: new Set(), // found in src/node_errors.h
+  noargs: new Set(), // subset of `js` which is constructed without arguments
+  // `set.manual` contains errors manually created from js and not defined in
```
It is not immediately clear if this comment is a continuation of the previous line or a new section. Maybe uppercase the previous (if not all) line and add a period or insert a line break before?

vsemozhetbyt commented on test/parallel/test-error-codes-impl-doc.js:
```diff
@@ -0,0 +1,156 @@
+  }
+}
+
+// Check that we have can initialize errors called without args
```
we have can -> we can?

vsemozhetbyt commented on test/parallel/test-error-codes-impl-doc.js:
```diff
@@ -0,0 +1,156 @@
+  } catch (e) {
+    ok = false;
+  }
+  check(ok, 'init', name, 'failed 0 args init, but is called with "()"');
```
This comment can be a bit confusing, as `0` can be read as an argument. Maybe `failed no args init`?

ChALkeR replied:
Will change to `failed init without args`.

vsemozhetbyt commented on test/parallel/test-error-codes-impl-doc.js:
```diff
@@ -0,0 +1,156 @@
+  check(ok, 'init', name, 'failed 0 args init, but is called with "()"');
+}
+
+// Check that we have implementation for all errors mentioned in docs
```
Add a period to separate from the next sentence?

ChALkeR commented:
@vsemozhetbyt Thanks, comments addressed!

ChALkeR commented:
Remaining issues of 10.x not yet adressed with pull requests:
```console
js: ERR_TLS_RENEGOTIATE - missing JS implementation (source)
js: ERR_MISSING_DYNAMIC_INTSTANTIATE_HOOK - missing JS implementation (source)
js: ERR_HTTP2_SETTINGS_CANCEL - missing JS implementation (source)
doc: ERR_TLS_RENEGOTIATE - missing documentation
doc: ERR_MISSING_DYNAMIC_INTSTANTIATE_HOOK - missing documentation
doc: ERR_HTTP2_SETTINGS_CANCEL - missing documentation
doc: ERR_TRANSFERRING_EXTERNALIZED_SHAREDARRAYBUFFER - missing documentation
doc: ERR_HTTP2_STREAM_CLOSED - missing documentation
There were 17 problems found
```

ChALkeR commented:
All the known issues with error codes were resolved (in other PRs), this should pass now.

ChALkeR commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/16210/

Trott commented:
@nodejs/testing

ChALkeR commented:
Ping everyone again ;-).
I updated this to ignore legacy error codes from the docs.

targos commented on test/parallel/test-error-codes-impl-doc.js:
```diff
@@ -0,0 +1,158 @@
+
+// Check that we can initialize errors called without args
+for (const name of set.noargs) {
+  if (!errors[name]) continue; // Already catched that above
```
catched -> caught

BridgeAR commented on test/parallel/test-error-codes-impl-doc.js:
```diff
@@ -0,0 +1,158 @@
+for (const file of fs.readdirSync(docdir)) {
+  if (!file.endsWith('.md')) continue;
+  let data = fs.readFileSync(path.join(docdir, file), 'utf-8');
+  if (file === 'errors.md') data = data.replace(/## Legacy [\s\S]*/, '');
```
Should `docdata` maybe just be read here? That way the file must only be read once instead of twice.

BridgeAR commented on test/parallel/test-error-codes-impl-doc.js:
```diff
@@ -0,0 +1,158 @@
+
+function addSource(source, type) {
+  // eslint-disable-next-line node-core/no-unescaped-regexp-dot
+  const re = /(.)?(ERR_[A-Z0-9_]+)(..)?/g;
```
If I understand the regexp correct the last part is there to potentially match a JS error called without arguments. If that's the case, could we just change this to: `(\(\))?`?

BridgeAR commented on test/parallel/test-error-codes-impl-doc.js:
```diff
@@ -0,0 +1,158 @@
+const report = process.argv[2] === '--report';
+const results = [];
+function check(ok, type, name, reason) {
+  const label = `${type}: ${name} - ${reason}`;
```
Since the last three argument combined become the label, wouldn't it be best to pass through the label directly?

BridgeAR commented on test/parallel/test-error-codes-impl-doc.js:
```diff
@@ -0,0 +1,158 @@
+const natives = process.binding('natives');
+
+// --report prints only the list of failed checks and does not assert
+const report = process.argv[2] === '--report';
```
I suggest to change that to always asserting but checking all entries first instead of asserting on the first failure.

That way the failures are obvious from the assertion and the user does not have to run the test with this flag (and it's not that obvious that this flag must be used by just taking a glimpse at this).

BridgeAR commented:
Ping @ChALkeR

fhinkel commented:
Closing this due to inactivity. Please reopen if needed.
